efi/preinstall/check_tcglog.go

```diff
@@ -0,0 +1,610 @@
+	switch t.phase {
+	case tcglogPhaseOSPresent:
+		// This is the OS-present phase, we no longer care for anything in the log once
+		// we're in this phase.
+		return t.phase, nil
+	case tcglogPhaseErr:
+		panic("not expecting any more events after returning an error")
+	}
+
+	// We're not in OS-present phase, so make sure the measurement is to a TCG defined PCR
+	if !internal_efi.IsTCGDefinedPCR(ev.PCRIndex) {
+		return 0, fmt.Errorf("measurements were made by firmware from pre-OS environment to non-TCG defined PCR %d", ev.PCRIndex)
+	}
+
+	// Handle any EV_SEPARATOR event
+	if ev.EventType == tcglog.EventTypeSeparator {
+		// Make sure it's the only one seen for this PCR
+		if _, exists := t.pcrSeparatorsSeen[ev.PCRIndex]; exists {
+			return 0, fmt.Errorf("more than one EV_SEPARATOR event exists for PCR %d", ev.PCRIndex)
+		}
+
+		// Make sure the separator doesn't signal an error condition. If an error condition occurs,
+		// the firmware should cap all PCRs with EV_SEPARATOR events that indicate that an error
+		// occurred.
+		data, ok := ev.Data.(*tcglog.SeparatorEventData)
+		if !ok {
+			// if it failed to decode then it's guaranteed to implement the error interface.
+			return 0, fmt.Errorf("invalid event data for EV_SEPARATOR event in PCR %d: %w", ev.PCRIndex, ev.Data.(error))
+		}
+		if data.IsError() {
+			// The EV_SEPARATOR event indicates that an error occurred.
+			return 0, fmt.Errorf("EV_SEPARATOR event for PCR %d indicates an error occurred (error code in log: %d)", ev.PCRIndex, binary.LittleEndian.Uint32(data.Bytes()))
+		}
+
+		t.pcrSeparatorsSeen[ev.PCRIndex] = struct{}{} // Mark the EV_SEPARATOR as seen for this PCR
+		t.numSeparatorsSeen += 1                      // Count the number of EV_SEPARATORs seen.
+	}
+
+	// Handle phase transitions
+	switch {
+	case t.phase == tcglogPhasePreOSBeforeMeasureSecureBootConfig:
+		switch {
+		case ev.EventType == tcglog.EventTypeSeparator:
+			// An EV_SEPARATOR during this phase is used to signal an error condition, but
+			// the data for the separator should have indicated the error, which is caught above.
+			// If we get here then this is a normal EV_SEPARATOR event.
+			return 0, fmt.Errorf("unexpected normal EV_SEPARATOR event in PCR %d", ev.PCRIndex)
+		case ev.PCRIndex == internal_efi.SecureBootPolicyPCR:
+			// A non EV_SEPARATOR event to PCR7 signals the beginning of the secure boot config measurements.
+			t.phase = tcglogPhasePreOSMeasuringSecureBootConfig
+		default:
+			// No change for any other event
+		}
+	case t.phase == tcglogPhasePreOSMeasuringSecureBootConfig:
+		switch {
+		case ev.PCRIndex != internal_efi.SecureBootPolicyPCR && ev.EventType == tcglog.EventTypeSeparator:
+			// An EV_SEPARATOR event in PCRs 0-6 after measuring the secure boot configuration begins the
+			// transition to OS-present, skipping the part of the pre-OS phase after the secure boot config
+			// has been measured.
+			t.phase = tcglogPhaseTransitioningToOSPresent
```

Comment:
this case is not tested

Reply:
Aha, I'll see if there's something I can do to test that. At the moment I'm just doing ad-hoc modifications to logs to break things.

Note that the next PR currently has quite a low coverage because it's got a switch statement with a lot of branches to cover all possible event types. We have tests for the most common types that we're likelly to see, but testing all of the others is going to require significant work in `internal/efitest` to customize the log and generate events with different types. I'm not sure whether that will be a blocker for the first iteration, especially as Oliver is keen to get this landed.